### ISSUE TITLE
[Question] Best practice to set Content-Security-Policy header on HttpServletResponse using salvation?

### ISSUE THREAD

**Issue by cnsgithub** [shapesecurity/salvation#195]
First of all, sorry about asking questions in the issue tracker, normally questions would go in the forum. However, I didn't find any forum related to salvation or shapesecurity. 

Also, I am missing javadoc and user guides for your nontheless promising project.

My goal is to dynamically set `Content-Security-Policy` headers, at the moment focusing on the `frame-ancestors` directive, which may contain allowed origins that users have configured. 

Right now I am doing some regex input validation to only allow origins that match `<scheme>://<host>[:<port>]`. 

However, when setting the hader on `HttpServletResponse` within a servlet, I am currently just doing some header escaping (removing carriage returns and line feeds).

There are two things I thought can be achieved with salvation:
1. Prevent injection of further directives via malformed/unvalidated user input
2. Do not override (instead merge) directives that have been set previously

As of the API usage, I don't understand the meaning of the `origin `parameter when bulding a `Policy` for example: `@Nonnull public static Policy parse(@Nonnull String sourceText, @Nonnull String origin)`

This is the way I go at the moment (just a snippet):
```java
protected void setFramingSecurityHeaders(HttpServletRequest request, HttpServletResponse response) {
        boolean denyFrames = true;
...
        List<String> allowedOrigins = ...; // WHITELIST
        String requestedOrigin = request.getParameter(RequestParam.ORIGIN.getName()); // TO PREVENT DISCLOSURE OF WHITELIST
        if (allowedOrigins.contains(requestedOrigin)) {
                        response.setHeader(com.google.common.net.HttpHeaders.X_FRAME_OPTIONS, "ALLOW-FROM " + OWASPEncoder.httpHeaderEncode(requestedOrigin)); // INTERNET EXPLORER
                        response.setHeader(com.google.common.net.HttpHeaders.CONTENT_SECURITY_POLICY, "frame-ancestors " + OWASPEncoder.httpHeaderEncode(requestedOrigin)); // FIXME: DO NOT OVERRIDE PREVIOUSLY SET HEADER, INSTEAD MERGE
                        denyFrames = false;
        }
...
        if (denyFrames) {
            response.setHeader(com.google.common.net.HttpHeaders.X_FRAME_OPTIONS, "DENY"); // INTERNET EXPLORER
            response.setHeader(com.google.common.net.HttpHeaders.CONTENT_SECURITY_POLICY, "frame-ancestors 'none'"); // FIXME: DO NOT OVERRIDE PREVIOUSLY SET HEADER, INSTEAD MERGE
       }
}
```

Could you please give me some hints on how to integrate salvation properly? Thanks.

**Comment by shekyan** [comment]
> Prevent injection of further directives via malformed/unvalidated user input
You can construct an empty policy and add whitelisted ancestor-sources:
```
       Origin origin = URI.parse("http://example.com");
        Policy p = new Policy(origin);
        Set<AncestorSource> ancestorSourceList = new HashSet<>();
        ancestorSourceList.add(new HostSource("https", "another-example.com", 443, null));
        ancestorSourceList.add(Self);
        p.addDirective(new FrameAncestorsDirective(ancestorSourceList));
        assertEquals("", p.show()); // frame-ancestors 'self' https://another-example.com
```

Salvation can intersect policies (that's the equivalent of having multiple CSP headers in the response):
```
       Origin origin = URI.parse("http://example.com");
       Policy p  = Parser.parse("frame-ancestors 'self' https://another-example.com", origin);
       Policy q = Parser.parse("frame-ancestors a b 'self'", origin);
        p.intersect(q);
        p.show(); // frame-ancestors 'self'
```
Origin is the origin of protected document and is needed to resolve relative URLs (in `report-uri`) and `'self'`, which might be needed during merging operations and optimisations. 

In general, I think README gives the idea on the usage, and tests give more specific examples. Feel free to email me you have more questions.

**Comment by cnsgithub** [comment]
Ok, what I am trying to do now is preserving previous `Content-Security-Policy `directives by using the `union `merge strategy. However, the [frame-ancestors ](url)diretive must be merged using the intersection strategy, otherwise something like `'none'` and `http://example.com` would be combined to `frame-ancestors http://example.com 'none'` which is invalid, see https://cspvalidator.org/#headerValue%5B%5D=frame-ancestors+http%3A%2F%2Fexample.com+'none'%3B+default-src&strategy=union

```java
    protected String mergeContentSecurityPolicyFrameAncestors(HttpServletRequest request, String allowedOrigin) {
        Origin origin = URI.parse(request.getRequestURL().toString());
        List<Policy> previousPolicies = new ArrayList<>();
        Iterator<String> previousHeaders = request.getHeaders(com.google.common.net.HttpHeaders.CONTENT_SECURITY_POLICY).asIterator();
        while (previousHeaders.hasNext()) {
            previousPolicies.add(Parser.parse(previousHeaders.next(), origin));
        }
        URI allowedOriginURI = URI.parse(allowedOrigin);
        Policy frameAncestorPolicy = new Policy(origin);
        frameAncestorPolicy.addDirective(new FrameAncestorsDirective(Collections.singleton(new HostSource(allowedOriginURI.scheme, allowedOriginURI.host, allowedOriginURI.port, null))));
        for (Policy previousPolicy : previousPolicies) {
            frameAncestorPolicy.union(previousPolicy);
            // FIXME use intersection for frame-ancestors directive 
        }
        return frameAncestorPolicy.show();
    }
```

How can I do that?

**Comment by shekyan** [comment]
You should go with intersection strategy for all of them. 
I would just intersect previous policies first, and intersect `frameAncestorPolicy ` with the result of intersection of previous policies. Then replace all CSPs with the new one.

**Comment by shekyan** [comment]
That's a bug in `unionDirective`, which can produce invalid source-expressions. We'll try to fix it asap.

**Comment by cnsgithub** [comment]
The bug in unionDirective seems to be fixed in 2.5.0 release. Thank you.

**Comment by shekyan** [comment]
Thanks for the feedback and pointing out the bug.
